queries/pic.py

```diff
@@ -0,0 +1,177 @@
+                "Sýndu mér mynd af Halldóri Laxness",
+                "Sýndu ljósmynd af Kára Stefánssyni",
+                "Sýndu mynd af Hörpunni",
+            )
+        )
+    )
+
+
+# This module wants to handle parse trees for queries
+HANDLE_TREE = True
+
+# The grammar nonterminals this module wants to handle
+QUERY_NONTERMINALS = {"QPic"}
+
+# The context-free grammar for the queries recognized by this plug-in module
+GRAMMAR = """
+
+Query →
+    QPic
+
+QPic →
+    QPicQuery '?'?
+
+QPicQuery →
+    QPicShowMePictureQuery | QPicWrongPictureQuery
+
+QPicShowMe →
+    "sýndu" | "getur" "þú" "sýnt" | "geturðu" "sýnt" | "viltu" "sýna" |
+    "nennirðu" "að" "sýna" | "nennir" "þú" "að" "sýna"
+
+QPicShowMePictureQuery →
+    QPicShowMe QPicMeOrUs? QPicPictureOrPhoto "af" QPicSubject
+
+QPicPictureOrPhoto →
+    "ljósmynd" | "mynd" | "ljósmyndir" | "myndir"
+
+QPicMeOrUs →
+    "mér" | "okkur"
+
+QPicSubject →
+    Nl_þgf
+
+QPicWrongPictureQuery →
+    "þetta" QPicIsWas QPicWrong QPicPictureOrPhoto
+
+QPicWrong →
+    "röng" | "vitlaus" | "ekki" "rétt"
+
+QPicIsWas →
+    "er" | "var"
+
+"""
+
+
+def QPicQuery(node: Node, params: QueryStateDict, result: Result) -> None:
+    # Set the query type
+    result.qtype = _PIC_QTYPE
+
+
+def _preprocess(s: str) -> str:
+    return s
+
+
+def QPicSubject(node: Node, params: QueryStateDict, result: Result) -> None:
+    n = _preprocess(result._text)
+    nom = NounPhrase(n).nominative or n
+    result.subject = nom
+    result.subject_þgf = result._text
+
+
+def QPicWrongPictureQuery(node: Node, params: QueryStateDict, result: Result) -> None:
+    result.wrong = True
+
+
+def _gen_pic_answer(result: Result, q: Query):
+    """Generate answer to query asking for a picture of something."""
+    subj = result["subject"]
+    subj_þgf = result["subject_þgf"]
+
+    # Look up picture using Google Images API
+    img: Optional[Img] = None
+    try:
+        img = get_image_url(subj)
+        if img and img.src:
+            # We found an image of the subject
+            answ = f"Hér er mynd sem ég fann af {icequote(subj_þgf)}"
+            q.set_answer(*gen_answer(answ))
+            q.set_image(img.src)
+            src: str = "Google Images"
+            o = urlparse(img.src)
+            if o and o.hostname:
+                src = o.hostname
+            q.set_source(src)
+            # q.set_expires(datetime.utcnow() + timedelta(hours=1))
+        else:
+            # No picture found
+            q.set_answer(*gen_answer(f"Engin mynd fannst af {icequote(subj_þgf)}"))
+    except Exception as e:
+        q.set_answer(*gen_answer(f"Ekki tókst að leita í myndagrunni"))
+
+    q.set_key(subj)
+    q.set_qtype(result.qtype)
+
+
+_WRONG_PIC_ANSWER = "Ég biðst afsökunar á því. Enginn er fullkominn."
+
+
+def _gen_wrong_pic_answer(result: Result, q: Query):
+    """Query states that previous picture was wrong."""
+    q.set_qtype(result.qtype)
+    q.set_answer(*gen_answer(_WRONG_PIC_ANSWER))
```

Comment:
We could actually use the query context to tell us which picture is being referred to, and mark it as wrong in the database (or collect this information in some other way for review or use).

Reply:
Good idea.